Add support for current azure environments

InKahootz opened the pull request:
Currently, the DNS validation for azure will produce errors when trying
to resolve cross-cloud requests. The AzureUSGoverenment specifically
forbids cross-cloud requests because it's a confidential environment.

This will add support for the four current public Azure environments:
AzureCloud, AzureChinaCloud, AzureUSGovernment, AzureGermanCloud.

Resolves: #1485

WouterTinus commented on src/plugin.validation.dns.azure/AzureEnvironment.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace PKISharp.WACS.Plugins.ValidationPlugins
+{
+    public class AzureEnvironment
```
This whole class seems like an overly complicated way to specify an URI. Being able to provide an URI would also enable people to use their private Azure services (i.e. Azure Stack). If your concern is that people won't know what their options are or what endpoint to specify, it's still possible to use the `ChooseFromList` functions to have the user select from a list of pre-defined strings, and perhaps a `null` or other 'semi-magic' answer could trigger a call to `RequestString` to get the custom URI.

InKahootz replied:
Do you happen to mean `ChooseFromMenu`? I'm not seeing any `ChooseFromList` in the `IInputService`.

I wasn't familiar with the FCLP library but I can work it out from existing code and implement a custom URI method.

InKahootz commented:
Sounds goods! I'll get on your suggestions right away!

InKahootz commented:
Okay, I've reworked it a bit. Simplified the URL class as well as adding error handling if nothing in the options exist that specify what environment to run under.

WouterTinus commented:
Thanks, I'll take another look soon.

WouterTinus commented on src/plugin.validation.dns.azure/Azure.cs:
```diff
@@ -16,17 +18,33 @@ internal class Azure : DnsValidation<Azure>
             _options = options;
             _domainParser = domainParser;
+
+            if (!AzureEnvironments.ResourceManagerUrls.TryGetValue(options.AzureEnvironment, out string endpoint))
```
This will still crash existing renewals, where `options.AzureEnvironment` is `null`.

WouterTinus commented on src/plugin.validation.dns.azure/AzureOptionsFactory.cs:
```diff
@@ -46,6 +68,7 @@ public override Task<AzureOptions> Default(Target target)
             var options = new AzureOptions
             {
                 UseMsi = az.AzureUseMsi,
+                AzureEnvironment = _arguments.TryGetRequiredArgument(nameof(az.AzureEnvironment), az.AzureEnvironment),
```
This will break existing workflows because `TryGetRequiredArgument` call will throw an exception when the argument you're asking for has not been provided. This should be an optional arguments and most preferably also be case insensitive, e.g. `AzureUSGovernment` and `AzureUsGovernment` should mean the same thing.

InKahootz commented:
Is it fine for the plugin to have a menu asking about which environment you want with the default to AzureCloud? Like it does in its current form?

WouterTinus commented:
> Is it fine for the plugin to have a menu asking about which environment you want with the default to AzureCloud? Like it does in its current form?

That is fine for interactive mode (the `Aquire` function) but in unattended mode (the `Default` function) we cannot ask for user input. 

In that case I think I would probably leave the environment unspecified, so that it works with the default at runtime, not the default at creation time.

LBegnaud commented:
@WouterTinus wouldn't you want to use whatever it used during initial runtime rather than potentially change every renewal?

WouterTinus commented:
That's a tradeoff between being able to easily change a setting for existing renewals without having to re-create them, and keeping the renewal configuration a stable as possible.

In similar scenario's I've chosen the former route, e.g. if you change the default Central Certificate Store path, all renewals that refer to the "default" change with it, which makes it easy to move it to another location. 

Maybe you are right that in this case the risk is greater because the likelihood of someone changing all of their DNS zones to another Azure cloud at once is pretty low.

WouterTinus commented:
I've merged this PR with the latest commits on 2.1.7 and addressed the BW compat issues. 

Not able to test though: one of the changes in 2.1.7 is the addition of proxy support for Azure, which requires us to provide the DnsManagementClient with our own (proxied) HttpClient. AFIAKT there is no constructor that accepts both an HttpClient *and* an base uri. So I hope I solved this by specifying the `BaseUri` property instead, but I don't know how the HttpClient and BaseUri interact with eachother.

So, please test :)

InKahootz commented:
👍 

Thank you! I’ve been busy with work. I’ll will try to test it in our environment this week.

Sent from my iPhone

> On Apr 19, 2020, at 2:43 PM, Wouter Tinus <notifications@github.com> wrote:
> 
> I've merged this PR with the latest commits on 2.1.7 and addressed the BW compat issues.
> 
> Not able to test though: one of the changes in 2.1.7 is the addition of proxy support for Azure, which requires us to provide the DnsManagementClient with our own (proxied) HttpClient. AFIAKT there is no constructor that accepts both an HttpClient and an base uri. So I hope I solved this by specifying the BaseUri property instead, but I don't know how the HttpClient and BaseUri interact with eachother.
> 
> So, please test :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

WouterTinus commented:
If you can test it in the next couple of days we can still included it with the 2.1.7 release, otherwise I'll have to wait until 2.1.8.

WouterTinus commented:
Has anyone been able to test this? The 2.1.8 release window is closing soon :)

InKahootz commented:
I finally tested the non-managed identity side. Seems you also need to provide specific environment settings for the ApplicationTokenProvider otherwise you get the CrossCloud errors that are usually seen. It was an easy fix. There was some discussion already on it here: https://github.com/Azure/azure-sdk-for-ruby/issues/850#issuecomment-319067598

InKahootz commented:
The managed service identity does get the credentials from the environment specific URL but there must be some additional permission setup required for it. Not quite sure how MSI works. I have always used the service provider tokens.

Here's the stack trace regardless. It fails when trying to read the dns zones so I would assume as long as permissions are set up correctly it would work.
```
 [EROR] Error preparing for challenge answer
Microsoft.Rest.Azure.CloudException: The client '991c804e-....' with object id '991c804e-....' does not have authorization to perform action 'Microsoft.Network/dnsZones/read' over scope '/subscriptions/<subscription-id_redacted>/resourceGroups/<my-dns-zone-rg_redacted>/providers/Microsoft.Network' or the scope is invalid. If access was recently granted, please refresh your credentials.
   at Microsoft.Azure.Management.Dns.ZonesOperations.ListByResourceGroupWithHttpMessagesAsync(String resourceGroupName, Nullable`1 top, Dictionary`2 customHeaders, CancellationToken cancellationToken)
   at Microsoft.Azure.Management.Dns.ZonesOperationsExtensions.ListByResourceGroupAsync(IZonesOperations operations, String resourceGroupName, Nullable`1 top, CancellationToken cancellationToken)
   at PKISharp.WACS.Plugins.ValidationPlugins.Dns.Azure.GetHostedZone(String recordName) in C:\Git\win-acme\src\plugin.validation.dns.azure\Azure.cs:line 130
   at PKISharp.WACS.Plugins.ValidationPlugins.Dns.Azure.CreateRecord(String recordName, String token) in C:\Git\win-acme\src\plugin.validation.dns.azure\Azure.cs:line 56
   at PKISharp.WACS.Plugins.ValidationPlugins.DnsValidation`1.PrepareChallenge()
   at PKISharp.WACS.Plugins.ValidationPlugins.Validation`1.PrepareChallenge(IChallengeValidationDetails challenge)
   at PKISharp.WACS.RenewalExecutor.HandleChallenge(ExecutionContext context, TargetPart targetPart, Authorization authorization)
```

WouterTinus commented:
Hi InKahootz, thanks for the updated code. I'm just about to release 2.1.9 tomorrow or the day after, so I hope you don't mind this will be the first major feature landing in 2.1.10 😄

InKahootz commented:
No worries! Glad to contribute to open source projects 😃

Sent from my iPhone

> On Jul 6, 2020, at 3:42 PM, Wouter Tinus <notifications@github.com> wrote:
> 
> Hi InKahootz, thanks for the updated code. I'm just about to release 2.1.9 tomorrow or the day after, so I hope you don't mind this will be the first major feature landing in 2.1.10 😄
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
